File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char

responseFor :: String -> String
responseFor xs
  | yell      = "Whoa, chill out!"
  | question  = "Sure."
  | silence   = "Fine. Be that way!"
  | otherwise = "Whatever."
  where
    yell     = (any isAlpha xs) && (map toUpper xs) == xs
    question = not (null stripped) && last stripped == '?' 
    silence  = all isSpace xs
    stripped = dropWhileEnd isSpace xs

File: go/bob/bob.go
package bob

import "strings"
import "unicode"

// Generates Bob's response to a remark
func Hey(remark string) string {
	switch {
	case isYell(remark) && isQuestion(remark):
		return "Calm down, I know what I'm doing!"
	case isYell(remark):
		return "Whoa, chill out!"
	case isQuestion(remark):
		return "Sure."
	case isSilence(remark):
		return "Fine. Be that way!"
	default:
		return "Whatever."
	}
}

func isQuestion(remark string) bool {
	return strings.HasSuffix(strings.TrimRight(remark, " "), "?")
}

func isYell(remark string) bool {
	return hasLetter(remark) && strings.ToUpper(remark) == remark
}

func hasLetter(remark string) bool {
	for _, r := range remark {
		if unicode.IsLetter(r) {
			return true
		}
	}
	return false
}

func isSilence(remark string) bool {
	for _, r := range remark {
		if !unicode.IsSpace(r) {
			return false
		}
	}
	return true
}
